ai/ai-react-app/src/components/Specific/ChatMessage.tsx

```diff
@@ -23,38 +37,176 @@ const getMessageText = (message: Content): string => {
+        });
+      }
+    }
+  });
+
+  if (segmentsToHighlight.length === 0) {
+    return [text];
+  }
+
+  // Sort segments by start index, then by end index
+  segmentsToHighlight.sort((a, b) => {
+    if (a.startIndex !== b.startIndex) {
+      return a.startIndex - b.startIndex;
+    }
+    return b.endIndex - a.endIndex; // Longer segments first
+  });
+
+  const outputNodes: React.ReactNode[] = [];
+  let lastIndexProcessed = 0;
+
+  segmentsToHighlight.forEach((seg, i) => {
+    // Add un-highlighted text before this segment
+    if (seg.startIndex > lastIndexProcessed) {
+      outputNodes.push(text.substring(lastIndexProcessed, seg.startIndex));
+    }
+
+    // Add the highlighted segment
+    // Ensure we don't re-highlight an already covered portion if a shorter segment comes later
+    const currentSegmentText = text.substring(seg.startIndex, seg.endIndex);
+    const tooltipText = seg.chunkIndices
+      .map((ci) => {
+        const chunk = chunks[ci - 1]; // ci is 1-based
+        return chunk.web?.title || chunk.web?.uri || `Source ${ci}`;
+      })
+      .join("; ");
+
+    outputNodes.push(
+      <span
+        key={`seg-${i}`}
+        className={styles.highlightedSegment}
+        title={`Sources: ${tooltipText}`}
+        data-source-indices={seg.chunkIndices.join(",")}
+      >
+        {currentSegmentText}
+        <sup className={styles.sourceSuperscript}>
+          [{seg.chunkIndices.join(",")}]
+        </sup>
+      </span>,
+    );
+    lastIndexProcessed = Math.max(lastIndexProcessed, seg.endIndex);
+  });
+
+  // Add any remaining un-highlighted text
+  if (lastIndexProcessed < text.length) {
+    outputNodes.push(text.substring(lastIndexProcessed));
+  }
+
+  return outputNodes;
+};
+
 /**
  * Renders a single chat message bubble, styled based on the message role ('user' or 'model').
  * It only renders messages that should be visible in the log (user messages, or model messages
  * containing text). Function role messages or model messages consisting only of function calls
  * are typically not rendered directly as chat bubbles.
  */
-const ChatMessage: React.FC<ChatMessageProps> = ({ message }) => {
+const ChatMessage: React.FC<ChatMessageProps> = ({
+  message,
+  groundingMetadata,
+}) => {
   const text = getMessageText(message);
   const isUser = message.role === "user";
   const isModel = message.role === "model";
 
-  // We render:
-  // 1. User messages (even if they only contain images/files, the 'user' role indicates an entry).
-  // 2. Model messages *only if* they contain actual text content.
-  // We *don't* render:
-  // 1. 'function' role messages (these represent execution results, not direct chat).
-  // 2. 'model' role messages that *only* contain function calls (these are instructions, not display text).
-  // 3. 'system' role messages (handled separately, not usually in chat history display).
-  const shouldRender = isUser || (isModel && text.trim() !== "");
+  const shouldRender =
```

Comment:
Should there still be an explanation here?

Reply:
Yes, I added the description back